docs/src/tutorials/getting_started/design_patterns_for_larger_models.jl

```diff
@@ -0,0 +1,610 @@
+end
+
+data
+
+# Then, we can re-write our `solve_knapsack` function to take our `KnapsackData`
+# as input:
+
+function solve_knapsack_2(data::KnapsackData)
+    model = Model(GLPK.Optimizer)
+    @variable(model, x[keys(data.objects)], Bin)
+    @objective(model, Max, sum(v.profit * x[k] for (k, v) in data.objects))
+    @constraint(
+        model,
+        sum(v.weight * x[k] for (k, v) in data.objects) <= data.capacity,
+    )
+    optimize!(model)
+    return value.(x)
+end
+
+solve_knapsack_2(data)
+
+# ## Read in data from files
+
+# Having a data structure is a good step. But it is still annoying that we have
+# to hard-code the data into Julia. A good next step is to separate the data
+# into an external file format; JSON is a common choice.
+
+import JSON
+
+function read_data(filename)
+    d = JSON.parsefile(filename)
+    return KnapsackData(
+        Dict(
+            k => KnapsackObject(v["profit"], v["weight"]) for
+            (k, v) in d["objects"]
+        ),
+        d["capacity"],
+    )
+end
+
+data = read_data(joinpath(@__DIR__, "data", "knapsack.json"))
+
+# ## Add options via if-else
+
+# At this point, we have data in a file format which we can load and solve a
+# single problem. For many users, this might be sufficient. However, at some
+# point you may be asked to add features like "but what if I want to take more
+# than one of a particular item?"
+
+# If this is the first time that you've been asked to add a feature, adding
+# options via `if-else` statements is a good approach. For example, we might
+# write:
+
+function solve_knapsack_3(data::KnapsackData; binary_knapsack::Bool)
+    model = Model(GLPK.Optimizer)
+    if binary_knapsack
+        @variable(model, x[keys(data.objects)], Bin)
+    else
+        @variable(model, x[keys(data.objects)] >= 0, Int)
+    end
+    @objective(model, Max, sum(v.profit * x[k] for (k, v) in data.objects))
+    @constraint(
+        model,
+        sum(v.weight * x[k] for (k, v) in data.objects) <= data.capacity,
+    )
+    optimize!(model)
+    return value.(x)
+end
+
+# Now we can solve the binary knapsack:
+
+solve_knapsack_3(data; binary_knapsack = true)
+
+# And an integer knapsack where we can take more than one copy of each item:
+
+solve_knapsack_3(data; binary_knapsack = false)
+
+# ## Add options via dispatch
+
+# If you get repeated requests to add different options, you'll quickly find
+# yourself in a mess of different flags and `if-else` statements. It's hard to
+# write, hard to read, and hard to ensure you haven't introduced any bugs.
+# A good solution is to use Julia's type dispatch. The easiest way to explain
+# this is by example.
+
+# First, start by defining a new abstract type, as well as new subtypes for each
+# of our options.
+
+abstract type AbstractKnapsack end
+
+struct BinaryKnapsack <: AbstractKnapsack end
+
+struct IntegerKnapsack <: AbstractKnapsack end
+
+# Then, we rewrite our `solve_knapsack` function to take a `config` argument,
```

Comment:
Calling it `config` was a little unclear to me at first: `config_type`, `knapsack_type`, `typed_knapsack`...?

Reply:
I changed `AbstractKnapsack` to `AbstractConfiguration`. Does that help?

Reply:
It does. But I did like the term `AbstractKnapsack` quite a bit.